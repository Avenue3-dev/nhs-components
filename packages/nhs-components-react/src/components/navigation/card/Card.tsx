import { HTMLProps } from 'react'
import clsx from 'clsx'
import { AsElementLink } from '@/types/link-like'

type Card = {
  Content: typeof Content
  Description: typeof Description
  Heading: typeof Heading
  Link: typeof Link
  Image: typeof Image
  Group: typeof Group
  GroupItem: typeof GroupItem
} & React.FC<CardProps>

type CardProps = {
  clickable?: boolean
} & HTMLProps<HTMLDivElement>

type GroupItemProps = {
  width?: GroupItemWidth
} & HTMLProps<HTMLLIElement>

type HeadingProps = {
  asElement?: 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6'
} & HTMLProps<HTMLHeadingElement>

export type GroupItemWidth = 'one-half' | 'one-third' | 'one-quarter'

const Content: React.FC<HTMLProps<HTMLDivElement>> = ({
  children,
  className,
  ...rest
}) => {
  return (
    <div className={clsx('nhsuk-card__content', className)} {...rest}>
      {children}
    </div>
  )
}

const Heading: React.FC<HeadingProps> = ({
  children,
  className,
  asElement: Component = 'h2',
  ...rest
}) => {
  return (
    <Component
      className={clsx('nhsuk-card__heading nhsuk-heading-m', className)}
      {...rest}
    >
      {children}
    </Component>
  )
}

const Link: React.FC<AsElementLink<HTMLAnchorElement>> = ({
  children,
  className,
  asElement: Component = 'a',
  ...rest
}) => {
  return (
    <Component className={clsx('nhsuk-card__link', className)} {...rest}>
      {children}
    </Component>
  )
}

const Description: React.FC<HTMLProps<HTMLParagraphElement>> = ({
  children,
  className,
  ...rest
}) => {
  return (
    <p className={clsx('nhsuk-card__description', className)} {...rest}>
      {children}
    </p>
  )
}

const Image: React.FC<HTMLProps<HTMLImageElement>> = ({
  className,
  ...rest
}) => {
  return <img className={clsx('nhsuk-card__img', className)} {...rest} />
}

const Group: React.FC<HTMLProps<HTMLUListElement>> = ({
  children,
  className,
  ...rest
}) => {
  return (
    <ul
      className={clsx('nhsuk-grid-row nhsuk-card-group', className)}
      {...rest}
    >
      {children}
    </ul>
  )
}

const GroupItem: React.FC<GroupItemProps> = ({
  children,
  className,
  width = 'one-half',
  ...rest
}) => {
  return (
    <li
      className={clsx(
        'nhsuk-card-group__item',
        `nhsuk-grid-column-${width}`,
        className,
      )}
      {...rest}
    >
      {children}
    </li>
  )
}

/**
 * Use a card to give users a brief summary of content or a task, often with a link to more detail. You can display a card alongside other cards to group related content or tasks.
 *
 * For more information on when to use this component, go to the [NHS Service Manuals website](https://service-manual.nhs.uk/design-system/components/card).
 *
 * @link https://service-manual.nhs.uk/design-system/components/card
 *
 * @example
 * ```tsx
 * <Card>
 *  <Card.Content>
 *   <Card.Heading>
 *    If you need help now, but itâ€™s not an emergency
 *   </Card.Heading>
 *   <Card.Description>
 *    Go to <a href="#site">111.nhs.uk</a> or <a href="#site">call 111</a>
 *   </Card.Description>
 *  </Card.Content>
 * </Card>
 * ```
 */
const Card: Card = ({ children, className, clickable, ...rest }) => {
  return (
    <div
      className={clsx(
        'nhsuk-card',
        { 'nhsuk-card--clickable': clickable },
        className,
      )}
      {...rest}
    >
      {children}
    </div>
  )
}

Card.Content = Content
Card.Description = Description
Card.Heading = Heading
Card.Link = Link
Card.Image = Image
Card.Group = Group
Card.GroupItem = GroupItem

Card.displayName = 'Card'
Content.displayName = 'Card.Content'
Description.displayName = 'Card.Description'
Heading.displayName = 'Card.Heading'
Link.displayName = 'Card.Link'
Image.displayName = 'Card.Image'
Group.displayName = 'Card.Group'
GroupItem.displayName = 'Card.GroupItem'

export default Card
