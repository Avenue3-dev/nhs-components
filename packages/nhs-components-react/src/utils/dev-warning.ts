const prod = process.env.NODE_ENV === 'production'

export const experimentalWarning = (
  component: 'tabs' | 'character-count',
): void => {
  if (prod) {
    return
  }

  console.warn(`ATTENTION
  This component is tagged as experimental and therefore could change without notice. See more details about this component at https://service-manual.nhs.uk/design-system/components/${component}
  `)
}
