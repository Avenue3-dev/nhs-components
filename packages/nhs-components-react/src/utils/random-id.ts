export const generateRandomString = (length = 5): string => {
  const rand = Math.random() + 1

  return rand.toString(36).substring(2, length + 2)
}

export const generateRandomName = (prefix?: string): string => {
  const randString = generateRandomString()

  return prefix ? `${prefix}_${randString}` : randString
}

export const generateRandomId = generateRandomName
