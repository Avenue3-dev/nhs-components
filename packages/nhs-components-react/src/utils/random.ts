export const generateRandomString = (length: number = 8): string => {
  return [...Array<string>(length)]
    .map(() => Math.random().toString(36)[2])
    .join('')
}
