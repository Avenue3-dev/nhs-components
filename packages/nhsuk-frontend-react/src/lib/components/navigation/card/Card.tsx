import { HTMLProps, createContext, useContext } from 'react'
import clsx from 'clsx'
import { AsElementLink } from '@/types/link-like'
import { AsElementHeadingProps } from '@/types/heading'
import { ChevronRightCircleIcon } from '@/icons'

type Card = {
  Content: typeof Content
  Description: typeof Description
  Heading: typeof Heading
  Link: typeof Link
  Image: typeof Image
  Group: typeof Group
  GroupItem: typeof GroupItem
} & React.FC<CardProps>

type CardProps = {
  clickable?: boolean
} & Partial<CardContextValue> &
  HTMLProps<HTMLDivElement>

type GroupItemProps = {
  width?: GroupItemWidth
} & HTMLProps<HTMLLIElement>

type HeadingProps = AsElementHeadingProps & HTMLProps<HTMLHeadingElement>

type CardContextValue = {
  withChevron: boolean
  primary: boolean
  secondary: boolean
}

export type GroupItemWidth = 'one-half' | 'one-third' | 'one-quarter'

const CardContext = createContext<CardContextValue>({
  withChevron: false,
  primary: false,
  secondary: false,
})

const Content: React.FC<HTMLProps<HTMLDivElement>> = ({
  children,
  className,
  ...rest
}) => {
  const { primary, secondary, withChevron } = useContext(CardContext)

  return (
    <div
      className={clsx(
        'nhsuk-card__content',
        {
          'nhsuk-card__content--primary': primary,
          'nhsuk-card__content--secondary': secondary && !primary,
        },
        className,
      )}
      {...rest}
    >
      {children}

      {withChevron && primary && <ChevronRightCircleIcon />}
    </div>
  )
}

const Heading: React.FC<HeadingProps> = ({
  children,
  className,
  asElement: Component = 'h2',
  ...rest
}) => {
  return (
    <Component
      className={clsx('nhsuk-card__heading nhsuk-heading-m', className)}
      {...rest}
    >
      {children}
    </Component>
  )
}

const Link: React.FC<AsElementLink<HTMLAnchorElement>> = ({
  children,
  className,
  asElement: Component = 'a',
  ...rest
}) => {
  return (
    <Component className={clsx('nhsuk-card__link', className)} {...rest}>
      {children}
    </Component>
  )
}

const Description: React.FC<HTMLProps<HTMLParagraphElement>> = ({
  children,
  className,
  ...rest
}) => {
  return (
    <p className={clsx('nhsuk-card__description', className)} {...rest}>
      {children}
    </p>
  )
}

const Image: React.FC<HTMLProps<HTMLImageElement>> = ({
  className,
  alt,
  ...rest
}) => {
  return <img className={clsx('nhsuk-card__img', className)} alt={alt} {...rest} />
}

const Group: React.FC<HTMLProps<HTMLUListElement>> = ({
  children,
  className,
  ...rest
}) => {
  return (
    <ul
      className={clsx('nhsuk-grid-row nhsuk-card-group', className)}
      {...rest}
    >
      {children}
    </ul>
  )
}

const GroupItem: React.FC<GroupItemProps> = ({
  children,
  className,
  width = 'one-half',
  ...rest
}) => {
  return (
    <li
      className={clsx(
        'nhsuk-card-group__item',
        `nhsuk-grid-column-${width}`,
        className,
      )}
      {...rest}
    >
      {children}
    </li>
  )
}

/**
 * Use a card to give users a brief summary of content or a task, often with a link to more detail. You can display a card alongside other cards to group related content or tasks.
 *
 * For more information on when to use this component, go to the [NHS Service Manuals website](https://service-manual.nhs.uk/design-system/components/card).
 *
 * @link https://service-manual.nhs.uk/design-system/components/card
 *
 * @example
 * ```tsx
 * <Card>
 *  <Card.Content>
 *   <Card.Heading>
 *    If you need help now, but itâ€™s not an emergency
 *   </Card.Heading>
 *   <Card.Description>
 *    Go to <a href="#site">111.nhs.uk</a> or <a href="#site">call 111</a>
 *   </Card.Description>
 *  </Card.Content>
 * </Card>
 * ```
 */
const Card: Card = ({
  children,
  className,
  clickable,
  withChevron = false,
  primary = false,
  secondary = false,
  ...rest
}) => {
  return (
    <div
      className={clsx(
        'nhsuk-card',
        {
          'nhsuk-card--clickable': clickable,
          'nhsuk-card--secondary': secondary && !primary,
        },
        className,
      )}
      {...rest}
    >
      <CardContext.Provider value={{ withChevron, primary, secondary }}>
        {children}
      </CardContext.Provider>
    </div>
  )
}

CardContext.displayName = 'CardContext'

Card.Content = Content
Card.Description = Description
Card.Heading = Heading
Card.Link = Link
Card.Image = Image
Card.Group = Group
Card.GroupItem = GroupItem

Card.displayName = 'Card'
Content.displayName = 'Card.Content'
Description.displayName = 'Card.Description'
Heading.displayName = 'Card.Heading'
Link.displayName = 'Card.Link'
Image.displayName = 'Card.Image'
Group.displayName = 'Card.Group'
GroupItem.displayName = 'Card.GroupItem'

export default Card
