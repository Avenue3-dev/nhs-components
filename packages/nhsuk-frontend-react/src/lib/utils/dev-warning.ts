const prod = process.env.NODE_ENV === 'production'

/**
 * Print a warning when using experimental components when not in production mode.
 *
 * @param component name of the experimental component.
 */
export const useExperimentalWarning = (
  component: 'tabs' | 'character-count',
): void => {
  useDevWarning(`This component is tagged as experimental and therefore could change without notice.

  See more details about this component at https://service-manual.nhs.uk/design-system/components/${component}
  `)
}

/**
 * Print a warning message when not in production mode.
 *
 * @param message message to print when not it production.
 */
export const useDevWarning = (message: string): void => {
  if (prod) {
    return
  }

  console.warn(message)
}
