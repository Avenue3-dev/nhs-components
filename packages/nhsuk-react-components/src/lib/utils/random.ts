export const generateRandomString = (length = 8): string => {
  return [...Array<string>(length)]
    .map(() => Math.random().toString(36)[2])
    .join('')
}
