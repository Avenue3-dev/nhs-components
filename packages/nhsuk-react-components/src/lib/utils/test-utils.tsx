/* eslint-disable import/export */
import {
  RenderOptions,
  RenderResult,
  cleanup,
  render,
} from '@testing-library/react'
import { MemoryRouter as Router } from 'react-router-dom'
import { afterEach } from 'vitest'

afterEach(() => {
  cleanup()
})

const customRender = (
  ui: React.ReactElement,
  options: RenderOptions = {},
): RenderResult => {
  return render(ui, {
    // wrap provider(s) here if needed
    wrapper: ({ children }) => <Router>{children}</Router>,
    ...options,
  })
}

export * from '@testing-library/react'
export { default as userEvent } from '@testing-library/user-event'
// override render export
export { customRender as render }
