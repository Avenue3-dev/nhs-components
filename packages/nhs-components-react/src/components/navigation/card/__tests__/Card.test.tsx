import { render, screen } from '@/utils/test-utils'
import { describe, expect, test } from 'vitest'
import Card from '..'
import { GroupItemWidth } from '../Card'

describe('Card', () => {
  test('should match the snapshot', () => {
    render(
      <Card data-testid="card">
        <Card.Content>
          <Card.Heading>Card heading</Card.Heading>
          <Card.Description>Card description</Card.Description>
        </Card.Content>
      </Card>,
    )
    expect(screen.getByTestId('card')).toMatchSnapshot()
  })

  test('should match the snapshot - clickable', () => {
    render(
      <Card data-testid="card" clickable>
        <Card.Content>
          <Card.Heading>
            <Card.Link href="#care-support">Card heading</Card.Link>
          </Card.Heading>
          <Card.Description>Card description</Card.Description>
        </Card.Content>
      </Card>,
    )
    expect(screen.getByTestId('card')).toMatchSnapshot()
  })

  test.each(['one-half', 'one-quarter', 'one-third', undefined] as GroupItemWidth[])(
    'should match the snapshot - width %s',
    (width) => {
      render(
        <Card.Group data-testid="card">
          <Card.GroupItem width={width}>
            <Card>
              <Card.Content>
                <Card.Heading>Card heading</Card.Heading>
                <Card.Description>Card description</Card.Description>
              </Card.Content>
            </Card>
          </Card.GroupItem>
        </Card.Group>,
      )
      expect(screen.getByTestId('card')).toMatchSnapshot()
    },
  )

  test('should match the snapshot - image and clickable', () => {
    render(
      <Card data-testid="card" clickable>
        <Card.Image
          src="https://assets.nhs.uk/prod/images/A_0218_exercise-main_FKW1X7.width-690.jpg"
          alt="A lady smiling"
        />
        <Card.Content>
          <Card.Heading>Card heading</Card.Heading>
          <Card.Description>Card description</Card.Description>
        </Card.Content>
      </Card>,
    )
    expect(screen.getByTestId('card')).toMatchSnapshot()
  })

  test('should match the snapshot - primary', () => {
    render(
      <Card data-testid="card" primary>
        <Card.Content>
          <Card.Heading>Card heading</Card.Heading>
          <Card.Description>Card description</Card.Description>
        </Card.Content>
      </Card>,
    )
    expect(screen.getByTestId('card')).toMatchSnapshot()
  })

  test('should match the snapshot - primary & withChevron', () => {
    render(
      <Card data-testid="card" primary withChevron>
        <Card.Content>
          <Card.Heading>Card heading</Card.Heading>
          <Card.Description>Card description</Card.Description>
        </Card.Content>
      </Card>,
    )
    expect(screen.getByTestId('card')).toMatchSnapshot()
  })

  test('should match the snapshot - secondary', () => {
    render(
      <Card data-testid="card" secondary>
        <Card.Content>
          <Card.Heading>Card heading</Card.Heading>
          <Card.Description>Card description</Card.Description>
        </Card.Content>
      </Card>,
    )
    expect(screen.getByTestId('card')).toMatchSnapshot()
  })
})
