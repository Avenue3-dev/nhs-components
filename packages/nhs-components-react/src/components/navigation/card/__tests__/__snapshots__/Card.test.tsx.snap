// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Card > should match the snapshot - clickable 1`] = `
<div
  class="nhsuk-card nhsuk-card--clickable"
  data-testid="card"
>
  <div
    class="nhsuk-card__content"
  >
    <h2
      class="nhsuk-card__heading nhsuk-heading-m"
    >
      <a
        class="nhsuk-card__link"
        href="#care-support"
      >
        Card heading
      </a>
    </h2>
    <p
      class="nhsuk-card__description"
    >
      Card description
    </p>
  </div>
</div>
`;

exports[`Card > should match the snapshot - image and clickable 1`] = `
<div
  class="nhsuk-card nhsuk-card--clickable"
  data-testid="card"
>
  <img
    alt="A lady smiling"
    class="nhsuk-card__img"
    src="https://assets.nhs.uk/prod/images/A_0218_exercise-main_FKW1X7.width-690.jpg"
  />
  <div
    class="nhsuk-card__content"
  >
    <h2
      class="nhsuk-card__heading nhsuk-heading-m"
    >
      Card heading
    </h2>
    <p
      class="nhsuk-card__description"
    >
      Card description
    </p>
  </div>
</div>
`;

exports[`Card > should match the snapshot - width one-half 1`] = `
<ul
  class="nhsuk-grid-row nhsuk-card-group"
  data-testid="card"
>
  <li
    class="nhsuk-card-group__item nhsuk-grid-column-one-half"
  >
    <div
      class="nhsuk-card"
    >
      <div
        class="nhsuk-card__content"
      >
        <h2
          class="nhsuk-card__heading nhsuk-heading-m"
        >
          Card heading
        </h2>
        <p
          class="nhsuk-card__description"
        >
          Card description
        </p>
      </div>
    </div>
  </li>
</ul>
`;

exports[`Card > should match the snapshot - width one-quarter 1`] = `
<ul
  class="nhsuk-grid-row nhsuk-card-group"
  data-testid="card"
>
  <li
    class="nhsuk-card-group__item nhsuk-grid-column-one-quarter"
  >
    <div
      class="nhsuk-card"
    >
      <div
        class="nhsuk-card__content"
      >
        <h2
          class="nhsuk-card__heading nhsuk-heading-m"
        >
          Card heading
        </h2>
        <p
          class="nhsuk-card__description"
        >
          Card description
        </p>
      </div>
    </div>
  </li>
</ul>
`;

exports[`Card > should match the snapshot - width one-third 1`] = `
<ul
  class="nhsuk-grid-row nhsuk-card-group"
  data-testid="card"
>
  <li
    class="nhsuk-card-group__item nhsuk-grid-column-one-third"
  >
    <div
      class="nhsuk-card"
    >
      <div
        class="nhsuk-card__content"
      >
        <h2
          class="nhsuk-card__heading nhsuk-heading-m"
        >
          Card heading
        </h2>
        <p
          class="nhsuk-card__description"
        >
          Card description
        </p>
      </div>
    </div>
  </li>
</ul>
`;

exports[`Card > should match the snapshot - width undefined 1`] = `
<ul
  class="nhsuk-grid-row nhsuk-card-group"
  data-testid="card"
>
  <li
    class="nhsuk-card-group__item nhsuk-grid-column-one-half"
  >
    <div
      class="nhsuk-card"
    >
      <div
        class="nhsuk-card__content"
      >
        <h2
          class="nhsuk-card__heading nhsuk-heading-m"
        >
          Card heading
        </h2>
        <p
          class="nhsuk-card__description"
        >
          Card description
        </p>
      </div>
    </div>
  </li>
</ul>
`;

exports[`Card > should match the snapshot 1`] = `
<div
  class="nhsuk-card"
  data-testid="card"
>
  <div
    class="nhsuk-card__content"
  >
    <h2
      class="nhsuk-card__heading nhsuk-heading-m"
    >
      Card heading
    </h2>
    <p
      class="nhsuk-card__description"
    >
      Card description
    </p>
  </div>
</div>
`;
